proposals/kapp/001-crd-upgrade-safety/README.md

```diff
@@ -0,0 +1,95 @@
+
+### Goals and Non-goals
+
+#### Goals
+- Providing `kapp` users a way to enable verification that an `update` operation on CRDs is safe and won't cause data loss or break clients
+
+#### Non-goals
+- Anything not pertaining to providing `kapp` users a way to ensure that `update` operations on CRDs are safe before applying changes
+
+### Specification
+A new preflight check will be added to `kapp` called `CRDUpgradeSafety`.
+
+#### `preflight` Package Changes
+The `CRDUpgradeSafety` preflight check is a very nuanced preflight check and different users will likely want to be able to configure this check a bit further to do things like:
+- Only issue warnings about potential problems
+- Fail closed/open
+
+In order to accomodate this configurability without introducing various versions of the same preflight check, the logic in https://github.com/carvel-dev/kapp/tree/develop/pkg/kapp/preflight will be updated to allow for configurable preflight checks.
+
+As an example, configuring a preflight check could look something like:
+```sh
+kapp deploy ... --preflight=Check=true:{"mode":"strict"}
+```
+
+To provide a high quality user experience, each configurable preflight check should have a sane default and not require a user to provide any configuration information.
+
+#### `CRDUpgradeSafety` Logic
+Updates to a CRD can be deemed unsafe if any of the following occur:
+- An existing version that is a stored version in the object store is removed
+- Scope of a CRD is changed from Namespace to Cluster (or vice versa)
+- Removing a field in an existing version
+- Adding a new required field
+- Updates to an existing field that makes validation more restrictive or would break existing versions/clients. Some brief examples: changing a field's type from string to int, increasing minimum value, removing an enum value.
+
+The most straightforward changes to check for are the first four and will have explicit checks in the implementation.
+
+Checking for breaking changes to an existing field is much more nuanced and is more difficult to catch all valid cases.
+It is proposed that an iterative approach is taken where the base implementation does not contain any specific checks against changed fields, but is written in a way that it is easy to add checks in the future.
+This does not mean that we shouldn't add any specific checks against changed fields before a release of `kapp` containing this preflight check, but rather focus on laying a foundation in which we can iteratively build on.
+Some potential field checks we could add after the base implementation is in place are:
+- Type is not changed
+- Enum value is not removed
+- Minimum value is not increased
+- Maximum value is not decreased
+
+To start with, there will be a couple configuration options for this preflight check:
```

Comment:
How would these configuration options be passed to kapp? It might help to provide some examples.

Reply:
https://github.com/carvel-dev/carvel/pull/729/files#diff-4a58b74b898751ee41d36dd947d166e4e1e47bbb62869512fb192f269857cb7aR49-R52

Reply:
Updated the config info in https://github.com/carvel-dev/carvel/pull/729/commits/5d56b1f645a139ac817062bb026c4ea77617e146 and it contains an example of what the config entries could look like. Marking as resolved.